javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+
+        @Override
+        public Tuple2<BitSet<T>, BitSet<T>> partition(Predicate<? super T> predicate) {
+            Objects.requireNonNull(predicate, "predicate is null");
+            return iterator().partition(predicate).map(this::createFromAll, this::createFromAll);
+        }
+
+        @Override
+        public BitSet<T> filter(Predicate<? super T> predicate) {
+            Objects.requireNonNull(predicate, "predicate is null");
+            final BitSet<T> bitSet = createFromAll(iterator().filter(predicate));
+            return (bitSet.length() == length()) ? this : bitSet;
+        }
+
+        @Override
+        public  <C> Map<C, BitSet<T>> groupBy(Function<? super T, ? extends C> classifier) {
+            Objects.requireNonNull(classifier, "classifier is null");
+            return iterator().groupBy(classifier).map((key, iterator) -> Tuple.of(key, createFromAll(iterator)));
+        }
+
+        @Override
+        public Comparator<T> comparator() {
+            return (t1, t2) -> Integer.compare(toInt.apply(t1), toInt.apply(t2));
+        }
+
+        @Override
+        public BitSet<T> takeWhile(Predicate<? super T> predicate) {
+            Objects.requireNonNull(predicate, "predicate is null");
+            final BitSet<T> result = createFromAll(iterator().takeWhile(predicate));
+            return (result.length() == length()) ? this : result;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public BitSet<T> addAll(Iterable<? extends T> elements) {
+            final Stream<Integer> source = Stream.ofAll(elements).map(toInt);
+            final long[] copy = copyExpand(1 + (source.max().getOrElse(0) >> ADDRESS_BITS_PER_WORD));
+            source.forEach(element -> {
+                if (element < 0) {
+                    throw new IllegalArgumentException("bitset element must be >= 0");
+                }
+                setElement(copy, element);
+            });
+            return fromBitMaskNoCopy(copy);
+        }
+
+        @Override
+        public boolean contains(T t) {
+            final int element = toInt.apply(t);
+            if (element < 0) {
+                throw new IllegalArgumentException("bitset element must be >= 0");
+            }
+            final int index = element >> ADDRESS_BITS_PER_WORD;
+            return (getWord(index) & (1L << element)) != 0;
+        }
+
+        @Override
+        public BitSet<T> init() {
+            if (isEmpty()) {
+                throw new UnsupportedOperationException("init of empty TreeSet");
+            } else {
+                final long last = getWord(getWordsNum() - 1);
+                final int element = BITS_PER_WORD * (getWordsNum() - 1) + BITS_PER_WORD - Long.numberOfLeadingZeros(last) - 1;
+                return remove(fromInt.apply(element));
+            }
+        }
+
+        @Override
+        public Iterator<T> iterator() {
+            return Stream.range(0, getWordsNum() << ADDRESS_BITS_PER_WORD)
```

Comment:
hmmm, this seams linear to the number of possible elements, and not to the existing ones.

Since this is an essential method, could we rather return a custom `Iterator` that uses `Integer.highestOneBit` to find the next bit in constant time for a block and jump through unset blocks the same way via `values[i] != 0` or something? It would still be linear, but should be a lot faster :/